### PR TITLE
Added feature of ScrollSpy's use of offset() or position() configurable

### PR DESCRIPTION
**Pull request by maximandrews**
Added feature requested in issue #16108

### CODE REVIEW

**Comment by cvrebert** [comment]
@twbs-savage Retry.

**Comment by cvrebert** [comment]
@twbs-savage retry

**Comment by cvrebert** [comment]
@twbs-savage Retry

**Comment by cvrebert** [comment]
@twbs-savage Retry
(Please!)

**Comment by cvrebert** [comment]
@twbs-savage Retry

**Comment by twbs-savage** [comment]
**Tests passed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **CONFIRMED**

Commit: 2791430a6a1340d3635d02152c93ddbf0624cafa
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/55843732

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

File: js/scrollspy.js [hunk 1]
```diff
@@ -40,14 +40,14 @@
 
   ScrollSpy.prototype.refresh = function () {
     var that          = this
-    var offsetMethod  = 'offset'
+    var offsetMethod  = this.options.offsetmethod === 'position' && 'position' || 'offset'
```
Comment (hnrch02) [inline js/scrollspy.js]: If you do it like this why not just rename the option and make it a boolean?

File: docs/_includes/js/scrollspy.html [hunk 2]
```diff
@@ -122,6 +122,12 @@ <h3 id="scrollspy-options">Options</h3>
          <td>10</td>
          <td>Pixels to offset from top when calculating position of scroll.</td>
        </tr>
+       <tr>
+         <td>offsetmethod</td>
```
Comment (hnrch02) [inline docs/_includes/js/scrollspy.html]: Should be camelCased.

File: js/scrollspy.js (hunk 1 above)
Comment (hnrch02) [inline js/scrollspy.js]: BTW, non-strict comparisons are used throughout Bootstrap's JS.

File: docs/_includes/js/scrollspy.html (hunk 2 above)
Comment (maxstorm) [inline docs/_includes/js/scrollspy.html]: Is that acceptable to have camelCased <code>data-</code> attributes? Keeping in mid your previous comment (boolean), may be it's better to change option name to <code>position</code>?

File: js/scrollspy.js [hunk 3]
```diff
@@ -40,14 +40,15 @@
 
   ScrollSpy.prototype.refresh = function () {
     var that          = this
-    var offsetMethod  = 'offset'
+    var forcePosition = this.options.position
+    var offsetMethod  = forcePosition && 'position' || 'offset'
```
Comment (hnrch02) [inline js/scrollspy.js]: Ternary would be clearer here IMO.

**Comment by twbs-savage** [comment]
**Tests failed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: 1bedf89e7a7cac5da394c44fc0a98ad4503174b7
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/55977397

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

File: js/scrollspy.js [hunk 4]
```diff
@@ -40,14 +40,15 @@
...
     var offsetBase    = 0
 
     this.offsets      = []
     this.targets      = []
     this.scrollHeight = this.getScrollHeight()
 
-    if (!$.isWindow(this.$scrollElement[0])) {
+    if (forcePosition === void 0 && !$.isWindow(this.$scrollElement[0])) {
```
Comment (hnrch02) [inline js/scrollspy.js]: We don't protect against overwritten undefined anywhere else, don't think we should start here.

Comment (hnrch02) [inline js/scrollspy.js]: Historically, we've also compared to undefined with `== null` where the explicit distinction isn't necessary.

**Comment by hnrch02** [comment]
@twbs-savage retry

**Comment by twbs-savage** [comment]
**Tests failed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: 1bedf89e7a7cac5da394c44fc0a98ad4503174b7
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/55983499

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

**Comment by twbs-savage** [comment]
**Tests failed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: 6c200bfa0a9306b978300ef6935e61e5e3c48c33
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/55983976

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

**Comment by twbs-savage** [comment]
**Tests passed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **CONFIRMED**

Commit: 238ba44c797b75ed321c85b6e08879b627d3e375
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/55984461

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

**Comment by mdo** [comment]
Probably punt on this and the source issue for v4?

**Comment by fat** [comment]
yep, agree with @mdo - adding to list for v4 to consider

**Pull request closed without merging**